optax/contrib/momo.py

```diff
@@ -0,0 +1,245 @@
+  effective step size is
+    `min(learning_rate, <adaptive term>)`
+
+  where the adaptive term is computed on the fly. 
+
+  Note that in `update_fn` you need to pass the latest (batch) loss value to
+    the argument `value`.
+
+  References:
+    [Schaipp et al., 2023](https://arxiv.org/abs/2305.07583)
+  Args:
+    learning_rate: User-specified learning rate. Recommended to be chosen
+      rather large, by default 1.0.
+    beta: Momentum coefficient (for EMA).
+    lb: Lower bound of the loss. Zero should be a good choice for many tasks.
+    weight_decay: Weight-decay parameter.
+
+  Returns:
+    A `GradientTransformation` object.
+  """
+  def init_fn(params: base.Params) -> MomoState:
+    exp_avg = tu.tree_map(lambda p: jnp.zeros(p.shape), params)
+    barf = 0
+    gamma = 0
+    count = jnp.zeros([], jnp.int32)
+    return MomoState(exp_avg, barf, gamma, count)
+
+  def update_fn(
+      updates: base.Updates,
+      state: MomoState,
+      params: Optional[base.Params],
+      value: Optional[Array] = None) -> tuple[base.Updates, MomoState]:
+    if params is None:
+      raise ValueError(base.NO_PARAMS_MSG)
+    if value is None:
+      raise ValueError("""You need to pass the latest loss value to Momo.
+                       Use `jax.value_and_grad` for this.""")
+    count = state.count
+    # initialize at first gradient, and loss
+    bt = cond(count == 0, lambda: 0., lambda: beta)
+    barf = bt*state.barf + (1-bt)*value
+    exp_avg = tu.tree_map(
+      lambda ea, g: bt*ea + (1-bt)*g,
+      state.exp_avg,
+      updates
+      )
+    gamma = bt*state.gamma + (1-bt)*tree_utils.tree_vdot(updates, params)
+    exp_avg_norm = tree_utils.tree_l2_norm(exp_avg,squared=True)
+    iprod = tree_utils.tree_vdot(exp_avg, params)
+    alpha = learning_rate(count) if callable(learning_rate) else learning_rate
+    t1 = jnp.maximum((1+alpha*weight_decay)*(
+                            barf - lb - gamma
+                            )  + iprod , 0)/(exp_avg_norm)
+    # if denom is zero, take no step
+    t1 = cond(exp_avg_norm <= jnp.finfo(float).eps,
+              lambda: 0.,
+              lambda: t1
+        )
+    tau = jnp.minimum(alpha, t1)
+    p_update = tu.tree_map(
+      lambda ea, p:
+      -(alpha*weight_decay)/(1+alpha*weight_decay)*p
+      - tau*ea,
+      exp_avg, params
+    )
+    new_state = MomoState(
+      exp_avg=exp_avg,
+      barf=barf,
+      gamma=gamma,
+      count=utils.safe_int32_increment(count)
+    )
+    return p_update, new_state
+
+  return base.GradientTransformationExtraArgs(init_fn, update_fn)
+
+class MomoAdamState(NamedTuple):
+  """State of the `GradientTransformation` returned by `momo_adam`."""
+  exp_avg: base.Updates
+  exp_avg_sq: base.Updates
+  barf: float
+  gamma: float
+  count: float
+
+
+def momo_adam(
+    learning_rate: base.ScalarOrSchedule = 1e-2,
+    betas: tuple[float, float] = (0.9, 0.999),
+    eps: float = 1e-8,
+    lb: float = 0.0,
```

Comment:
how about calling this `f_min` as with `polyak_sgd` ? https://optax.readthedocs.io/en/latest/api/optimizers.html#optax.polyak_sgd

Reply:
I would prefer the term lower bound, as this is closer to how we describe this quantity in the paper (strictly speaking, you don't need the optimal value for deriving MoMo, but only a lower bound). But I see your point of consistent naming...

Reply:
ok, but let's at least name it `lower_bound` ? lb is not very descriptive

Reply:
sure, can do that. Just fyi that I now added also an option where the lower bounded is estimated on the fly, which then will make some variable names a bit lengthy :) but at least the function argument should have a descriptive name

Reply:
yes exactly, having descriptive names in the function signature is the most important. Its fine if private variables have more cyptic names.